massa-execution-worker/src/worker.rs

```diff
@@ -0,0 +1,548 @@
+    /// Saturates down to 0 on negative durations.
+    /// Note that config.cursor_delay is taken into account.
+    fn get_time_until_next_active_slot(&self) -> MassaTime {
+        // get the timestamp of the slot after the current last active one
+        let next_slot = self
+            .last_active_slot
+            .get_next_slot(self.config.thread_count)
+            .expect("active slot overflow in VM");
+        let next_timestmap = get_block_slot_timestamp(
+            self.config.thread_count,
+            self.config.t0,
+            self.config.genesis_timestamp,
+            next_slot,
+        )
+        .expect("could not compute block timestmap in VM");
+
+        // get the current timestamp minus the cursor delay
+        let end_time = MassaTime::compensated_now(self.config.clock_compensation)
+            .expect("could not get current time in VM")
+            .saturating_sub(self.config.cursor_delay);
+
+        // compute the time difference, saturating down to zero
+        next_timestmap.saturating_sub(end_time)
+    }
+
+    /// Tells the execution state about the new sequence of active slots.
+    /// If some slots already executed in a speculative way changed,
+    /// or if one of their have predecessor slots changed,
+    /// the execution state will truncate the execution output history
+    /// to remove all out-of-date execution outputs.
+    /// Speculative execution will then resume from the point of truncation.
+    pub fn truncate_execution_history(&mut self) {
+        // acquire write access to execution state
+        let mut exec_state = self
+            .execution_state
+            .write()
+            .expect("could not lock execution state for writing");
+
+        // tells the execution state to truncate its execution output history
+        // given the new list of active slots
+        exec_state.truncate_history(&self.active_slots);
+    }
+
+    /// Executes a read-only request, and asynchronously returns the result once finished.
+    ///
+    /// # Arguments
+    /// * req: read-only execution request parameters
+    /// * resp_tx: MPSC sender through which the execution output is sent when the execution is over
+    fn execute_readonly_request(
+        &self,
+        req: ReadOnlyExecutionRequest,
+        resp_tx: mpsc::Sender<Result<ExecutionOutput, ExecutionError>>,
+    ) {
+        // acquire read access to execution state and execute the read-only request
+        let outcome = self
+            .execution_state
+            .read()
+            .expect("could not lock execution state for reading")
+            .execute_readonly_request(req);
+
+        // send the execution output through resp_tx
+        if resp_tx.send(outcome).is_err() {
+            debug!("could not send execute_readonly_request response: response channel died");
+        }
+    }
+
+    /// Main loop of the executin worker
+    pub fn main_loop(&mut self) {
```

Comment:
1. and 3.: I'm not entirely sure how this can be achieved properly, given the fact that the loops can also happen if there is no input. For example we loop whenever we process a single active block, in order to allow new requests to arrive and potentially cancel part of the previous one if it is still running (eg. history rewrite). As for the "check if something was added during the iteration", that's there to keep the input data unlocked during loop iteration to allow new data in even if the current loop iteration takes time.
Maybe I misunderstand your point. Can you can attempt a code snippet ?

2. this stems from the idea that in the future, consensus will not manage time anymore. And to reduce coupling and avoidable message passing, each component that needs time info can compute its own get_latest_slot.

Reply:
Ok sure, for both points I'll make a change on a fork of your branch to show what I mean.

Reply:
Ok I got it. Implemented !